### PR TITLE
[sdk-gen, dotnet/go] Fixes SDK-generation when referencing shared types in config variables

### LINKED ISSUE #15751: Non-primitive types can generate an invalid dotnet SDK
### What happened?

We generate an invalid dotnet SDK if we have any non-primitive types shared between provider config and resources.

I'm not sure if this is a go-provider bug or an issue with dotnet codegen, although our other SDKs don't seem to have issue with it. Edit: Fraser confirms it looks specific to dotnet.

### Example

```bash
cd pulumi-go-provider/examples/credentials
```

Apply the patch below. This adds a new type used in the provider's config as well as a resource.

```diff
diff --git a/examples/credentials/main.go b/examples/credentials/main.go
index c377e4d..39c6a83 100644
--- a/examples/credentials/main.go
+++ b/examples/credentials/main.go
@@ -30,12 +30,18 @@ func provider() p.Provider {
        })
 }

+type Shared struct {
+       Foo string `pulumi:"foo,optional"`
+}
+
 type Config struct {
        User     string   `pulumi:"user"`
        Password string   `pulumi:"password,optional" provider:"secret"`
        HashKind HashKind `pulumi:"hash"`

        hashedPassword string
+
+       Shared Shared `pulumi:"shared"`
 }

 type HashKind string
@@ -61,7 +67,6 @@ func (c *Config) Annotate(a infer.Annotator) {
        a.Describe(&c.Password, "The password. It is very secret.")
        a.Describe(&c.HashKind, `The (entirely uncryptographic) hash function used to encode the "password".`)
        a.SetDefault(&c.Password, "", "FOO")
-       a.SetDefault(&c.HashKind, HashAdler)
 }

 var _ = (infer.CustomConfigure)((*Config)(nil))
@@ -82,7 +87,9 @@ func (c *Config) Configure(ctx p.Context) error {
 }

 type User struct{}
-type UserArgs struct{}
+type UserArgs struct {
+       Shared Shared `pulumi:"shared"`
+}
 type UserState struct {
        Name     string `pulumi:"name"`
        Password string `pulumi:"password"`
```

(The `SetDefault` removal was needed to get `main` building. Filed a related issue [here](https://github.com/pulumi/pulumi-go-provider/issues/202).)

Generate an SDK and attempt to build it:

```bash
go build -o pulumi-resource-credentials . \
    && pulumi package gen-sdk ./pulumi-resource-credentials --language dotnet \
    && echo 0.0.1 > sdk/dotnet/version.txt \
    && cd sdk/dotnet \
    && dotnet build .
```

Get the following error:

> pulumi-go-provider/examples/credentials/sdk/dotnet/Config/Config.cs(56,23): error CS0246: The type or namespace name 'Types' could not be found (are you missing a using directive or an assembly reference?)

### Output of `pulumi about`

❯ pulumi about
CLI
Version      3.111.1
Go Version   go1.22.1
Go Compiler  gc

Host
OS       darwin
Version  14.1
Arch     arm64

Backend
Name           pulumi.com
URL            https://app.pulumi.com/bryce-pulumi-corp
User           bryce-pulumi-corp
Organizations  bryce-pulumi-corp, pulumi
Token type     personal

### Additional context

`Config.cs` is incorrectly expecting `Types.Shared`:

```cs
private static readonly __Value<Types.Shared?> _shared = new __Value<Types.Shared?>(() => __config.GetObject<Types.Shared>("shared"));
public static Types.Shared? Shared
{
    get => _shared.Get();
    set => _shared.Set(value);
}
```

`Provider.cs` and `User.cs` both correctly expect `Inputs.SharedArgs`:
```cs
[Input("shared", required: true, json: true)]
public Input<Inputs.SharedArgs> Shared { get; set; } = null!;
```

The schema is:
```json
{
  "name": "credentials",
  "version": "0.1.0",
  "meta": {
    "moduleFormat": "(.*)"
  },
  "config": {
    "variables": {
      "hash": {
        "$ref": "#/types/credentials:index:HashKind",
        "description": "The (entirely uncryptographic) hash function used to encode the \"password\"."
      },
      "password": {
        "type": "string",
        "description": "The password. It is very secret.",
        "default": "",
        "defaultInfo": {
          "environment": [
            "FOO"
          ]
        },
        "secret": true
      },
      "shared": {
        "$ref": "#/types/credentials:index:Shared"
      },
      "user": {
        "type": "string",
        "description": "The username. Its important but not secret."
      }
    },
    "defaults": [
      "hash",
      "shared",
      "user"
    ]
  },
  "types": {
    "credentials:index:HashKind": {
      "type": "string",
      "enum": [
        {
          "description": "Adler32 implements the Adler-32 checksum.",
          "value": "Adler32"
        },
        {
          "description": "CRC32 implements the 32-bit cyclic redundancy check, or CRC-32, checksum.",
          "value": "CRC32"
        }
      ]
    },
    "credentials:index:Shared": {
      "properties": {
        "foo": {
          "type": "string"
        }
      },
      "type": "object"
    }
  },
  "provider": {
    "properties": {
      "password": {
        "type": "string",
        "description": "The password. It is very secret.",
        "default": "",
        "defaultInfo": {
          "environment": [
            "FOO"
          ]
        },
        "secret": true
      },
      "user": {
        "type": "string",
        "description": "The username. Its important but not secret."
      }
    },
    "type": "object",
    "required": [
      "user"
    ],
    "inputProperties": {
      "hash": {
        "$ref": "#/types/credentials:index:HashKind",
        "description": "The (entirely uncryptographic) hash function used to encode the \"password\"."
      },
      "password": {
        "type": "string",
        "description": "The password. It is very secret.",
        "default": "",
        "defaultInfo": {
          "environment": [
            "FOO"
          ]
        },
        "secret": true
      },
      "shared": {
        "$ref": "#/types/credentials:index:Shared"
      },
      "user": {
        "type": "string",
        "description": "The username. Its important but not secret."
      }
    },
    "requiredInputs": [
      "hash",
      "shared",
      "user"
    ]
  },
  "resources": {
    "credentials:index:User": {
      "properties": {
        "name": {
          "type": "string"
        },
        "password": {
          "type": "string"
        }
      },
      "type": "object",
      "required": [
        "name",
        "password"
      ],
      "inputProperties": {
        "shared": {
          "$ref": "#/types/credentials:index:Shared"
        }
      },
      "requiredInputs": [
        "shared"
      ]
    }
  }
}
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
# Description

Fixes #15751 
- In .NET, non-primitive types such as objects used in config variables will be generated anyways under the `Types` namespace as plain-shaped objects inside of the `Config` class. This is because the types from `Outputs` or `Inputs` cannot be used directly from config. Also no longer marking objects used from config variables as "outputs" (doesn't make any sense)

- In Go, disable adding imports into the `config.go` based on used shared types because these are not used/referenced in the config module anywhere anyways.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/codegen/dotnet/gen.go [hunk 1]
```diff
@@ -2391,10 +2394,6 @@ func generateModuleContextMap(tool string, pkg *schema.Package) (map[string]*mod
 		cfg.namespaceName = fmt.Sprintf("%s.%s", cfg.RootNamespace(), namespaceName(infos[pkg].Namespaces, pkg.Name))
 	}
 
-	visitObjectTypes(pkg.Config, func(t *schema.ObjectType) {
-		getModFromToken(t.Token, pkg.Reference()).details(t).outputType = true
-	})
-
```
Comment (Zaid-Ajaj) [inline pkg/codegen/dotnet/gen.go:2397]: This part here is potentially a breaking change ⚠️ object types referenced from config variables were marked as outputs and were generated inside the `Outputs` namespace. This change removes them from there

**Comment by blampe** [comment]
Thanks for jumping on this @Zaid-Ajaj!

Out of curiosity, do we run `dotnet build` on the generated output to confirm things stay valid? When I looked at this briefly last week, the Configstation example on master didn't build for me, but I wasn't sure if that was an error on my part or if its output were legitimately broken. The former is most likely but the latter would be concerning so I just want to check if we have a sanity check in place.

**Comment by Zaid-Ajaj** [comment]
> Out of curiosity, do we run dotnet build on the generated output to confirm things stay valid?

@blampe Yes we do, the test schema added in this PR generates the .NET SDK then builds it to ensure type resolution is correct. 

> When I looked at this briefly last week, the Configstation example on master didn't build for me

I actually tried to fix that example with this PR but this specific schema of `configstation` actually fails due to a different issue with namespace resolution with the root namespace set to `Configstation.Pulumi` + package name being `Configstation`. I indent to fix the `configstation` example as soon as this lands ✅

**Comment by blampe** [comment]
I just ran this against the provider where I saw this issue and I can confirm it fixes the broken SDK 👍

**Review by Frassle — APPROVED** [review]

**Pull request merged**